### PR TITLE
[distributed] Add fake and hybrid topologies

### PR DESCRIPTION
**Pull request by sanketpurandare**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.18.0) (oldest at bottom):
* #4543
* #4542
* #4541
* #4665
* #4623
* #4559
* __->__ #4652
* #4591
* #4592
* #4540
* #4539
* #4643

Summary:
Add two deliberately separate fake-process-group modes for validating large logical training topologies without treating fake communication as a substitute for a real distributed run.

`fake_backend` represents one selected rank of the complete logical mesh with fake process groups on every axis. It is the inexpensive first gate for configuration parsing, model and pipeline partitioning, sharding metadata, prepared-weight ownership, tensor shapes, and PyTorch-managed rank-local memory. It performs no peer data transfer and therefore cannot validate distributed numerics, collective behavior, or performance.

`real_pp_fake_spmd_backend` launches one physical process per PP rank, creates a real NCCL pipeline group, and keeps DP, TP, CP, and EP process groups fake. It is the next gate when the behavior under test depends on pipeline sends and receives, receive-buffer lifetime, schedule ordering, or whole-step CUDA-graph capture, while the complete logical SPMD world would be too large to allocate physically. It still cannot prove SPMD collective correctness or communication performance.

Communication initialization now returns one typed `DistributedTopology` instead of constructing a fake `DeviceMesh` and mutating it afterward. `ParallelDims` consumes the logical world size and any selected real axis groups once during mesh construction. There is no mesh override, fallback from one mode to the other, or inferred coordinate that can hide an invalid launch.

For logical world size `NGPU` and PP degree `P`, the non-PP logical world has size `NGPU / P` and the represented logical rank is:

`FAKE_PP_RANK * (NGPU / P) + FAKE_SPMD_RANK`

Pure fake PP requires both explicit coordinates. In hybrid mode, `torchrun`'s physical `RANK` is the PP coordinate, physical `WORLD_SIZE` must equal `P`, and `FAKE_SPMD_RANK` selects one common flattened SPMD coordinate for the PP line. Initialization validates divisibility, coordinate ranges, world size, rendezvous results, and real-group construction.

`docs/debugging.md` is the user-facing contract. It explains when to choose each mode, every environment variable (`NGPU`, `FAKE_PP_RANK`, `FAKE_SPMD_RANK`, and the standard torchrun variables), exact pure-fake and hybrid launch commands, represented and unrepresented memory, failure interpretation, and escalation to real distributed validation. The accompanying skill turns that contract into a repeatable logical-rank memory-snapshot workflow.

Test Plan:
- `python -m pytest -q tests/unit_tests/cpu/test_distributed_utils.py tests/unit_tests/cpu/test_parallel_dims.py tests/unit_tests/cpu/test_pipeline_parallel.py tests/unit_tests/cpu/test_config_manager.py tests/unit_tests/cpu/test_dist_moe.py` (145 passed)
- Exact documented one-GPU pure-fake PP2/DP2 command completed one step as logical rank 3
- Exact documented two-GPU real-PP/fake-SPMD PP2/DP2 command completed one step with real NCCL PP traffic and clean teardown on both ranks
- Scoped pre-commit hooks, codespell, and link checking passed; targeted Pyrefly reported 0 errors. The project-wide hook retains one unrelated nullable video-duration baseline error.

### CODE REVIEW

File: docs/debugging.md [hunk 1]
```diff
@@ -60,31 +60,142 @@ This will print a structured configuration to `stdout`, allowing you to verify t
...
+required for SPMD communication, numerical, and performance evidence.
+
+### Logical topology
+
+`NGPU` is always the logical world size, not necessarily the number of launched
+processes. For a pipeline degree `P`, each pipeline coordinate contains
+`NGPU / P` flattened SPMD coordinates. The selected logical global rank is:
+
+```text
+logical_rank = FAKE_PP_RANK * (NGPU / P) + FAKE_SPMD_RANK
```
Comment (tianyu-l) [inline docs/debugging.md:83]: this is assuming PP is always the outermost mesh -- is it always true?

File: torchtitan/config/configs.py [hunk 2]
```diff
@@ -401,13 +401,21 @@ class CommConfig:
     save_traces_file_prefix: str = "rank_"
     """Flight recorder trace files prefix"""
 
-    mode: Literal["default", "fake_backend"] = "default"
-    """
-    Communication mode for distributed training.
+    mode: Literal["default", "fake_backend", "real_pp_fake_spmd_backend"] = "default"
```
Comment (tianyu-l) [inline torchtitan/config/configs.py:404]:
```suggestion
    backend: Literal["default", "fake", "real_pp_fake_spmd"] = "default"
```

File: docs/debugging.md [hunk 3]
```diff
@@ -60,31 +60,142 @@ This will print a structured configuration to `stdout`, allowing you to verify t
...
 
-Set `COMM_MODE="fake_backend"` to validate your configuration, model setup, and rank-0 program logic without requiring full multi-GPU distributed execution. To inspect a nonzero rank under `torchrun`, see [Distributed Breakpoints and LOG_RANK](#distributed-breakpoints-and-log_rank).
+TorchTitan has two fake-process-group modes because they answer different
+debugging questions. They intentionally do not share a fallback path:
 
-```bash
-NGPU=32 COMM_MODE="fake_backend" ./run_train.sh
+| Mode | Physical processes | Real communication | Use it to validate |
+| --- | --- | --- | --- |
+| `fake_backend` | One | None | Configuration, logical mesh construction, rank-local model ownership, tensor shapes, and PyTorch-managed memory for one selected logical rank. |
```
Comment (tianyu-l) [inline docs/debugging.md:68]: does it work with PP, couldn't find any documentation.

If it doesn't, I don't see a reason to keep two different backends for them.

File: torchtitan/distributed/parallel_dims.py [hunk 4]
```diff
@@ -249,8 +275,29 @@ def unflatten_mesh(
             self._global_meshes["spmd_sparse_for_fwdbwd"] = full_sparse_mesh[
                 "dp_replicate", "efsdp", "ep"
             ]
+        pp_mesh = dataloading_mesh["pp"]
+        pp_group = next(
+            (
+                group
+                for axis, group in self._real_axis_groups
+                if axis == MeshAxisName.PP
```
Comment (tianyu-l) [inline torchtitan/distributed/parallel_dims.py:283]: confusing -- is this PP specific? What if I specify DP in `_real_axis_groups` what happens?

File: docs/debugging.md [hunk 5]
```diff
@@ -60,31 +60,142 @@ This will print a structured configuration to `stdout`, allowing you to verify t
...
+PP2 on one physical GPU. It validates that rank's stage, shards, prepared
+weights, pipeline metadata, and memory ownership without creating NCCL process
+groups or transferring peer data.
 
-**Example use case:**
 ```bash
-# Validate a 128-GPU configuration on a single GPU
-NGPU=128 COMM_MODE="fake_backend" MODULE=llama3 CONFIG=llama3_70b ./run_train.sh
+NGPU=4 \
+FAKE_PP_RANK=1 \
```
Comment (tianyu-l) [inline docs/debugging.md:113]: should we set different `FAKE_PP_RANK` on different device? How?
